scm/build/Dockerfile

```diff
@@ -1,24 +1,139 @@
-FROM alpine:latest
+#
+# Stage: Build container. Used to build all binaries, once
+#
+FROM alpine:latest AS build-deps
 MAINTAINER Levente Kale <levente.kale@nokia.com>
 
-ENV GOPATH=/usr/local/go
-ENV PATH $GOPATH/bin:/usr/local/go/bin:$PATH
+ENV GOPATH=/go
+ENV GOROOT=/usr/local/go
+ENV PATH ${GOPATH}/bin:/usr/local/go/bin:$PATH
 ENV GO_VERSION 1.12.9
-
-COPY build.sh /build.sh
+ENV GOOS linux
+ENV CGO_ENABLED 0
+ENV GOARCH amd64
 
 RUN apk add --no-cache ca-certificates \
  && apk update --no-cache \
  && apk upgrade --no-cache \
  && apk add --no-cache make gcc musl-dev glide git bash curl tar \
  && mkdir -p ${GOPATH} \
+ && mkdir -p ${GOROOT} \
  && mkdir -p ${GOPATH}/bin \
  && mkdir -p ${GOPATH}/src \
- && curl -fsSL -k https://dl.google.com/go/go${GO_VERSION}.src.tar.gz | tar zx --strip-components=1 -C ${GOPATH} \
- && cd ${GOPATH}/src/ \
+ && curl -fsSL -k https://dl.google.com/go/go${GO_VERSION}.src.tar.gz | tar zx --strip-components=1 -C ${GOROOT} \
+ && cd ${GOROOT}/src/ \
  && ./make.bash \
  && rm -rf /var/cache/apk/* \
  && rm -rf /var/lib/apt/lists/* \
  && rm -rf /tmp/*
 
-ENTRYPOINT /build.sh
+RUN mkdir -p ${GOPATH}/src/github.com/nokia/danm
+COPY glide.yaml glide.lock ${GOPATH}/src/github.com/nokia/danm/
+COPY scm/build/build-deps.sh /
+
+WORKDIR ${GOPATH}/src/github.com/nokia/danm
+
+RUN /build-deps.sh
+
+
+#
+# Stage: Build actual binaries. This builder is used to build all DANM
+# related binaries.
+#
+# It is separated from the "build-deps" stage because mostly because
+# building the dependencies takes quite a lot of time and they (presumably)
+# don't change often -- so this could be factored out into a reusable
+# base image.
+#
+FROM build-deps AS builder
+
+COPY / ${GOPATH}/src/github.com/nokia/danm/
+
+ARG COMMIT_HASH
+ARG LATEST_TAG
+
+RUN scm/build/build.sh
+
+
+#
+# Stage: Netwatcher
+#
+FROM alpine:latest AS netwatcher
+MAINTAINER Levente Kale <levente.kale@nokia.com>
+
+COPY --from=builder /go/bin/netwatcher /usr/local/bin/netwatcher
+
+RUN apk add --no-cache --virtual .tools curl libcap iputils  \
+ && adduser -u 147 -D -H -s /sbin/nologin danm \
+ && chown root:danm /usr/local/bin/netwatcher \
+ && chmod 750 /usr/local/bin/netwatcher \
+ && setcap cap_sys_ptrace,cap_sys_admin,cap_net_admin=eip /usr/local/bin/netwatcher \
+ && setcap cap_net_raw=eip /usr/sbin/arping \
```

Comment:
It was long ago when danm was relying on arping binary, is there any other component that depends on it?
How about doing `setcap` for needed binaries in `builder` container? Would it be set for the copied file?

Reply:
Will remove the arping.

Unfortunately, no, xattrs (including what's set by setcap) are not preserved in a multistage build, it seems

Quick test:
```
% cat Dockerfile
FROM alpine:latest as stage1
RUN apk add --no-cache libcap attr \
 && touch test && echo "S1 BEFORE SET:" && getcap -v test && attr -l test \
 && setcap cap_net_admin=eip test  && echo "S1 AFTER SET:" && getcap -v test && attr -l test

FROM alpine:latest as stage2
COPY --from=stage1 test test
RUN apk add --no-cache libcap attr \
 && echo "STAGE2:" && getcap -v test && attr -l test
```

output (abbreviated):

```
S1 BEFORE SET:
test
S1 AFTER SET:
test = cap_net_admin+eip
Attribute "capability" has a 20 byte value for test

STAGE2:
test
```

https://github.com/moby/moby/issues/35699

Would have been nice, though...

Reply:
Oh dear! That means my spider-sense was warning me in the past. I did not know I was coming accross this issue while testing of PR #190 Migrate to go modules on a live system. It a pity. Thanks for investigating it. 👍